# Build Model Deployment For Car Price Predictions

url = [link](https://youtu.be/iRCaMnR_bpA)

In [1]:
import pandas as pd
import numpy as np

In [2]:
url = "https://raw.githubusercontent.com/rajtilakls2510/car_price_predictor/master/quikr_car.csv"

df = pd.read_csv(url)

In [3]:
df.head()

name   company  year          Price  \
0    Hyundai Santro Xing XO eRLX Euro III   Hyundai  2007         80,000   
1                 Mahindra Jeep CL550 MDI  Mahindra  2006       4,25,000   
2              Maruti Suzuki Alto 800 Vxi    Maruti  2018  Ask For Price   
3  Hyundai Grand i10 Magna 1.2 Kappa VTVT   Hyundai  2014       3,25,000   
4        Ford EcoSport Titanium 1.5L TDCi      Ford  2014       5,75,000   

   kms_driven fuel_type  
0  45,000 kms    Petrol  
1      40 kms    Diesel  
2  22,000 kms    Petrol  
3  28,000 kms    Petrol  
4  36,000 kms    Diesel

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 892 entries, 0 to 891
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   name        892 non-null    object
 1   company     892 non-null    object
 2   year        892 non-null    object
 3   Price       892 non-null    object
 4   kms_driven  840 non-null    object
 5   fuel_type   837 non-null    object
dtypes: object(6)
memory usage: 41.9+ KB


In [5]:
df.shape

(892, 6)

In [6]:
df.isna().sum()

name           0
company        0
year           0
Price          0
kms_driven    52
fuel_type     55
dtype: int64

In [7]:
df['year'].unique()

array(['2007', '2006', '2018', '2014', '2015', '2012', '2013', '2016',
       '2010', '2017', '2008', '2011', '2019', '2009', '2005', '2000',
       '...', '150k', 'TOUR', '2003', 'r 15', '2004', 'Zest', '/-Rs',
       'sale', '1995', 'ara)', '2002', 'SELL', '2001', 'tion', 'odel',
       '2 bs', 'arry', 'Eon', 'o...', 'ture', 'emi', 'car', 'able', 'no.',
       'd...', 'SALE', 'digo', 'sell', 'd Ex', 'n...', 'e...', 'D...',
       ', Ac', 'go .', 'k...', 'o c4', 'zire', 'cent', 'Sumo', 'cab',
       't xe', 'EV2', 'r...', 'zest'], dtype=object)

As we can see above, that years has more than integer value so it need converted into integer

In [8]:
df['Price'].unique()

array(['80,000', '4,25,000', 'Ask For Price', '3,25,000', '5,75,000',
       '1,75,000', '1,90,000', '8,30,000', '2,50,000', '1,82,000',
       '3,15,000', '4,15,000', '3,20,000', '10,00,000', '5,00,000',
       '3,50,000', '1,60,000', '3,10,000', '75,000', '1,00,000',
       '2,90,000', '95,000', '1,80,000', '3,85,000', '1,05,000',
       '6,50,000', '6,89,999', '4,48,000', '5,49,000', '5,01,000',
       '4,89,999', '2,80,000', '3,49,999', '2,84,999', '3,45,000',
       '4,99,999', '2,35,000', '2,49,999', '14,75,000', '3,95,000',
       '2,20,000', '1,70,000', '85,000', '2,00,000', '5,70,000',
       '1,10,000', '4,48,999', '18,91,111', '1,59,500', '3,44,999',
       '4,49,999', '8,65,000', '6,99,000', '3,75,000', '2,24,999',
       '12,00,000', '1,95,000', '3,51,000', '2,40,000', '90,000',
       '1,55,000', '6,00,000', '1,89,500', '2,10,000', '3,90,000',
       '1,35,000', '16,00,000', '7,01,000', '2,65,000', '5,25,000',
       '3,72,000', '6,35,000', '5,50,000', '4,85,000', '3,29,5

In price we sometimes seeing the string value like *Ask For Price*

Also in price it separated with 'commas' so it need to be proceded

For my challenge I want to change it into Rupiah

In [9]:
df['kms_driven'].unique()

array(['45,000 kms', '40 kms', '22,000 kms', '28,000 kms', '36,000 kms',
       '59,000 kms', '41,000 kms', '25,000 kms', '24,530 kms',
       '60,000 kms', '30,000 kms', '32,000 kms', '48,660 kms',
       '4,000 kms', '16,934 kms', '43,000 kms', '35,550 kms',
       '39,522 kms', '39,000 kms', '55,000 kms', '72,000 kms',
       '15,975 kms', '70,000 kms', '23,452 kms', '35,522 kms',
       '48,508 kms', '15,487 kms', '82,000 kms', '20,000 kms',
       '68,000 kms', '38,000 kms', '27,000 kms', '33,000 kms',
       '46,000 kms', '16,000 kms', '47,000 kms', '35,000 kms',
       '30,874 kms', '15,000 kms', '29,685 kms', '1,30,000 kms',
       '19,000 kms', nan, '54,000 kms', '13,000 kms', '38,200 kms',
       '50,000 kms', '13,500 kms', '3,600 kms', '45,863 kms',
       '60,500 kms', '12,500 kms', '18,000 kms', '13,349 kms',
       '29,000 kms', '44,000 kms', '42,000 kms', '14,000 kms',
       '49,000 kms', '36,200 kms', '51,000 kms', '1,04,000 kms',
       '33,333 kms', '33,600 kms', '5,

In kms driven we can see commas and 'kms' string, and also there is some nan values

In [10]:
df['fuel_type'].unique()

array(['Petrol', 'Diesel', nan, 'LPG'], dtype=object)

There is 'nan' value in fuel_type so it need to be converted

In [11]:
df.head()

name   company  year          Price  \
0    Hyundai Santro Xing XO eRLX Euro III   Hyundai  2007         80,000   
1                 Mahindra Jeep CL550 MDI  Mahindra  2006       4,25,000   
2              Maruti Suzuki Alto 800 Vxi    Maruti  2018  Ask For Price   
3  Hyundai Grand i10 Magna 1.2 Kappa VTVT   Hyundai  2014       3,25,000   
4        Ford EcoSport Titanium 1.5L TDCi      Ford  2014       5,75,000   

   kms_driven fuel_type  
0  45,000 kms    Petrol  
1      40 kms    Diesel  
2  22,000 kms    Petrol  
3  28,000 kms    Petrol  
4  36,000 kms    Diesel

We can see that names of car is too long, and somehow it useless, 

So it will be converted, let's check the length of the name

In [12]:
length_count = []

for x in df['name']:
    length_count.append(len(x.split()))

set(length_count)

{1, 2, 3, 4, 5, 6, 7, 8, 9}

There are 9 uniques values, it mean the maximum word is 9 and the minimum is 1

In [13]:
import collections

In [14]:
counter = collections.Counter(length_count)

In [15]:
# df_counter = pd.DataFrame(counter.items(), columns=['Value', 'Occurence'])
df_counter = pd.DataFrame({'Value': counter.keys(), 'Occurence': counter.values()})
df_counter.sort_values(by=['Value'])

Value  Occurence
8      1          7
3      2        101
5      3        130
1      4        195
2      5        212
4      6        121
0      7         86
6      8         37
7      9          3

As you can see I will use three, because there are about 200+ data have names with length three

## Cleaning dataset

In [16]:
df_backup = df.copy()

### Year Cleansing

In [17]:
df['year']

0      2007
1      2006
2      2018
3      2014
4      2014
       ... 
887    zest
888    2018
889    2013
890    2014
891    2014
Name: year, Length: 892, dtype: object

In [18]:
df['year'].str.isnumeric()

0       True
1       True
2       True
3       True
4       True
       ...  
887    False
888     True
889     True
890     True
891     True
Name: year, Length: 892, dtype: bool

In [19]:
df = df[df['year'].str.isnumeric()]

In [20]:
df['year']

0      2007
1      2006
2      2018
3      2014
4      2014
       ... 
886    2009
888    2018
889    2013
890    2014
891    2014
Name: year, Length: 842, dtype: object

In [21]:
# Turn into integer

df['year'] = df['year'].astype(int)

In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 842 entries, 0 to 891
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   name        842 non-null    object
 1   company     842 non-null    object
 2   year        842 non-null    int32 
 3   Price       842 non-null    object
 4   kms_driven  840 non-null    object
 5   fuel_type   837 non-null    object
dtypes: int32(1), object(5)
memory usage: 42.8+ KB


### Price Cleansing

In [23]:
df['Price'].unique()

array(['80,000', '4,25,000', 'Ask For Price', '3,25,000', '5,75,000',
       '1,75,000', '1,90,000', '8,30,000', '2,50,000', '1,82,000',
       '3,15,000', '4,15,000', '3,20,000', '10,00,000', '5,00,000',
       '3,50,000', '1,60,000', '3,10,000', '75,000', '1,00,000',
       '2,90,000', '95,000', '1,80,000', '3,85,000', '1,05,000',
       '6,50,000', '6,89,999', '4,48,000', '5,49,000', '5,01,000',
       '4,89,999', '2,80,000', '3,49,999', '2,84,999', '3,45,000',
       '4,99,999', '2,35,000', '2,49,999', '14,75,000', '3,95,000',
       '2,20,000', '1,70,000', '85,000', '2,00,000', '5,70,000',
       '1,10,000', '4,48,999', '18,91,111', '1,59,500', '3,44,999',
       '4,49,999', '8,65,000', '6,99,000', '3,75,000', '2,24,999',
       '12,00,000', '1,95,000', '3,51,000', '2,40,000', '90,000',
       '1,55,000', '6,00,000', '1,89,500', '2,10,000', '3,90,000',
       '1,35,000', '16,00,000', '7,01,000', '2,65,000', '5,25,000',
       '3,72,000', '6,35,000', '5,50,000', '4,85,000', '3,29,5

In [24]:
# Remove 'Ask For Price'

df = df[df['Price'] != 'Ask For Price']

In [25]:
df['Price'].unique()

array(['80,000', '4,25,000', '3,25,000', '5,75,000', '1,75,000',
       '1,90,000', '8,30,000', '2,50,000', '1,82,000', '3,15,000',
       '4,15,000', '3,20,000', '10,00,000', '5,00,000', '3,50,000',
       '1,60,000', '3,10,000', '75,000', '1,00,000', '2,90,000', '95,000',
       '1,80,000', '3,85,000', '1,05,000', '6,50,000', '6,89,999',
       '4,48,000', '5,49,000', '5,01,000', '4,89,999', '2,80,000',
       '3,49,999', '2,84,999', '3,45,000', '4,99,999', '2,35,000',
       '2,49,999', '14,75,000', '3,95,000', '2,20,000', '1,70,000',
       '85,000', '2,00,000', '5,70,000', '1,10,000', '4,48,999',
       '18,91,111', '1,59,500', '3,44,999', '4,49,999', '8,65,000',
       '6,99,000', '3,75,000', '2,24,999', '12,00,000', '1,95,000',
       '3,51,000', '2,40,000', '90,000', '1,55,000', '6,00,000',
       '1,89,500', '2,10,000', '3,90,000', '1,35,000', '16,00,000',
       '7,01,000', '2,65,000', '5,25,000', '3,72,000', '6,35,000',
       '5,50,000', '4,85,000', '3,29,500', '2,51,111', 

In [26]:
df['Price'].head()

0      80,000
1    4,25,000
3    3,25,000
4    5,75,000
6    1,75,000
Name: Price, dtype: object

In [27]:
# Remove commas

df['Price'] = df['Price'].str.replace(',', '').astype(int)

In [28]:
# Turn Rupee into Rupiah
# 1 Rupee = 194.23 Rupiah (13/06/2021)

df['Price'] = df['Price'] * 194.23

In [29]:
df['Price'] = df['Price'].astype(int)

In [30]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 819 entries, 0 to 891
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   name        819 non-null    object
 1   company     819 non-null    object
 2   year        819 non-null    int32 
 3   Price       819 non-null    int32 
 4   kms_driven  819 non-null    object
 5   fuel_type   816 non-null    object
dtypes: int32(2), object(4)
memory usage: 38.4+ KB


### KMS Driven

In [31]:
df['kms_driven'].head()

0    45,000 kms
1        40 kms
3    28,000 kms
4    36,000 kms
6    41,000 kms
Name: kms_driven, dtype: object

In [32]:
df['kms_driven'] = df['kms_driven'].str.split(' ').str.get(0).str.replace(',', '')

In [33]:
df['kms_driven']

0       45000
1          40
3       28000
4       36000
6       41000
        ...  
886    132000
888     27000
889     40000
890    Petrol
891    Petrol
Name: kms_driven, Length: 819, dtype: object

In [34]:
# Because there still string, let's turn it

df = df[df['kms_driven'].str.isnumeric()]

In [35]:
df['kms_driven'] = df['kms_driven'].astype(int)

In [36]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 817 entries, 0 to 889
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   name        817 non-null    object
 1   company     817 non-null    object
 2   year        817 non-null    int32 
 3   Price       817 non-null    int32 
 4   kms_driven  817 non-null    int32 
 5   fuel_type   816 non-null    object
dtypes: int32(3), object(3)
memory usage: 35.1+ KB


In [37]:
# Finding nan values

df[df['fuel_type'].isna()]

name company  year     Price  kms_driven fuel_type
132  Toyota Corolla  Toyota  2009  53413250       26000       NaN

In [38]:
df = df[~df['fuel_type'].isna()]

### Car Names

In [39]:
df['name'] = df['name'].str.split(' ').str.slice(0,3).str.join(' ')

In [40]:
df

name   company  year      Price  kms_driven fuel_type
0       Hyundai Santro Xing   Hyundai  2007   15538400       45000    Petrol
1       Mahindra Jeep CL550  Mahindra  2006   82547750          40    Diesel
3         Hyundai Grand i10   Hyundai  2014   63124750       28000    Petrol
4    Ford EcoSport Titanium      Ford  2014  111682250       36000    Diesel
6                 Ford Figo      Ford  2012   33990250       41000    Diesel
..                      ...       ...   ...        ...         ...       ...
883      Maruti Suzuki Ritz    Maruti  2011   52442100       50000    Petrol
885          Tata Indica V2      Tata  2009   21365300       30000    Diesel
886    Toyota Corolla Altis    Toyota  2009   58269000      132000    Petrol
888            Tata Zest XM      Tata  2018   50499800       27000    Diesel
889      Mahindra Quanto C8  Mahindra  2013   75749700       40000    Diesel

[816 rows x 6 columns]

As you can see we only have 816 rows but the index is 889

So it need to be reset first

In [41]:
df.reset_index(drop=True, inplace=True)

In [42]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 816 entries, 0 to 815
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   name        816 non-null    object
 1   company     816 non-null    object
 2   year        816 non-null    int32 
 3   Price       816 non-null    int32 
 4   kms_driven  816 non-null    int32 
 5   fuel_type   816 non-null    object
dtypes: int32(3), object(3)
memory usage: 28.8+ KB


In [43]:
df.describe()

year         Price     kms_driven
count   816.000000  8.160000e+02     816.000000
mean   2012.444853  7.996791e+07   46275.531863
std       4.002992  9.229507e+07   34297.428044
min    1995.000000  5.826900e+06       0.000000
25%    2010.000000  3.399025e+07   27000.000000
50%    2013.000000  5.826880e+07   41000.000000
75%    2015.000000  9.541549e+07   56818.500000
max    2019.000000  1.650956e+09  400000.000000

From the description we can know that the max is 1.650956e+09, this can be outliers

In [44]:
df[df['Price'] > 8e8]

name   company  year       Price  kms_driven fuel_type
534  Mahindra XUV500 W6  Mahindra  2014  1650955582       45000    Diesel

In [45]:
df = df[df['Price'] < 8e8].reset_index(drop=True)

In [46]:
df

name   company  year      Price  kms_driven fuel_type
0       Hyundai Santro Xing   Hyundai  2007   15538400       45000    Petrol
1       Mahindra Jeep CL550  Mahindra  2006   82547750          40    Diesel
2         Hyundai Grand i10   Hyundai  2014   63124750       28000    Petrol
3    Ford EcoSport Titanium      Ford  2014  111682250       36000    Diesel
4                 Ford Figo      Ford  2012   33990250       41000    Diesel
..                      ...       ...   ...        ...         ...       ...
810      Maruti Suzuki Ritz    Maruti  2011   52442100       50000    Petrol
811          Tata Indica V2      Tata  2009   21365300       30000    Diesel
812    Toyota Corolla Altis    Toyota  2009   58269000      132000    Petrol
813            Tata Zest XM      Tata  2018   50499800       27000    Diesel
814      Mahindra Quanto C8  Mahindra  2013   75749700       40000    Diesel

[815 rows x 6 columns]

In [47]:
df.to_csv('cleaned_car.csv')

## Modeling

In [48]:
X = df.drop(columns='Price')
y = df['Price']

In [49]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25)

In [52]:
from sklearn.compose import make_column_transformer
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import OneHotEncoder
from sklearn.pipeline import make_pipeline
from sklearn.metrics import r2_score


In [51]:
ohe = OneHotEncoder()
ohe.fit(X[['name', 'company', 'fuel_type']])

OneHotEncoder()

In [58]:
ohe.categories_

[array(['Audi A3 Cabriolet', 'Audi A4 1.8', 'Audi A4 2.0', 'Audi A6 2.0',
        'Audi A8', 'Audi Q3 2.0', 'Audi Q5 2.0', 'Audi Q7', 'BMW 3 Series',
        'BMW 5 Series', 'BMW 7 Series', 'BMW X1', 'BMW X1 sDrive20d',
        'BMW X1 xDrive20d', 'Chevrolet Beat', 'Chevrolet Beat Diesel',
        'Chevrolet Beat LS', 'Chevrolet Beat LT', 'Chevrolet Beat PS',
        'Chevrolet Cruze LTZ', 'Chevrolet Enjoy', 'Chevrolet Enjoy 1.4',
        'Chevrolet Sail 1.2', 'Chevrolet Sail UVA', 'Chevrolet Spark',
        'Chevrolet Spark 1.0', 'Chevrolet Spark LS', 'Chevrolet Spark LT',
        'Chevrolet Tavera LS', 'Chevrolet Tavera Neo', 'Datsun GO T',
        'Datsun Go Plus', 'Datsun Redi GO', 'Fiat Linea Emotion',
        'Fiat Petra ELX', 'Fiat Punto Emotion', 'Force Motors Force',
        'Force Motors One', 'Ford EcoSport', 'Ford EcoSport Ambiente',
        'Ford EcoSport Titanium', 'Ford EcoSport Trend',
        'Ford Endeavor 4x4', 'Ford Fiesta', 'Ford Fiesta SXi', 'Ford Figo',
        '

In [59]:
columns_trans = make_column_transformer((OneHotEncoder(categories=ohe.categories_), ['name', 'company', 'fuel_type']),
                                        remainder='passthrough')

In [61]:
linreg = LinearRegression()

In [62]:
pipe = make_pipeline(columns_trans, linreg)

In [63]:
pipe.fit(X_train, y_train)

Pipeline(steps=[('columntransformer',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('onehotencoder',
                                                  OneHotEncoder(categories=[array(['Audi A3 Cabriolet', 'Audi A4 1.8', 'Audi A4 2.0', 'Audi A6 2.0',
       'Audi A8', 'Audi Q3 2.0', 'Audi Q5 2.0', 'Audi Q7', 'BMW 3 Series',
       'BMW 5 Series', 'BMW 7 Series', 'BMW X1', 'BMW X1 sDrive20d',
       'BMW X1 xDrive20d', 'Chevrolet Beat', 'Chevrolet Beat...
                                                                            array(['Audi', 'BMW', 'Chevrolet', 'Datsun', 'Fiat', 'Force', 'Ford',
       'Hindustan', 'Honda', 'Hyundai', 'Jaguar', 'Jeep', 'Land',
       'Mahindra', 'Maruti', 'Mercedes', 'Mini', 'Mitsubishi', 'Nissan',
       'Renault', 'Skoda', 'Tata', 'Toyota', 'Volkswagen', 'Volvo'],
      dtype=object),
                                                                            array(['Diesel', 'LPG', 'Pe

In [64]:
y_pred = pipe.predict(X_test)

In [65]:
y_pred

array([ 3.85645494e+07,  1.06850161e+08,  4.57626795e+07,  5.21348915e+07,
        7.74623154e+07,  8.99447315e+07,  4.76062911e+07,  1.31260238e+08,
        2.55910386e+07,  8.97873971e+07,  1.93389758e+07,  5.06078674e+07,
        1.14020475e+08,  2.58414802e+07,  7.94069066e+06,  3.78910181e+07,
        1.18139996e+08,  5.17038111e+07,  1.22544961e+08,  4.27589522e+07,
        1.22932817e+08,  6.14047572e+07,  5.63294999e+07,  1.03129207e+08,
        5.32416804e+07,  7.79640222e+07,  1.00142889e+08, -1.72107404e+07,
        7.66128337e+07,  3.57104674e+07,  5.92294365e+07,  1.90314151e+07,
        4.78656521e+07,  6.39600845e+07,  4.14682668e+07,  5.92384854e+07,
        4.57626795e+07,  1.15363517e+08,  1.63715628e+08,  6.19183538e+07,
        5.16691534e+07,  6.87975121e+07,  5.45028000e+07,  1.15055309e+08,
        1.95952064e+07,  1.04438017e+08,  1.27094577e+08,  2.06189053e+08,
        8.50947079e+07,  6.62296586e+07,  5.07326019e+07,  2.23871247e+07,
        6.50324784e+07,  

In [66]:
r2_score(y_test, y_pred)

0.6877811387181065

To understand from where we got the number of 0.68, see this

In [ ]:
for i in range(10):
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25)
    linreg = LinearRegression()
    pipe = make_pipeline(columns_trans, linreg)
    pipe.fit(X_train, y_train)
    y_pred = pipe.predict(X_test)
    print(i, r2_score(y_test, y_pred))